File: app/src/main/java/com/ramez/shopp/Adapter/RecipeAdapter.java
package com.ramez.shopp.Adapter;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ramez.shopp.ApiHandler.DataFeacher;
import com.ramez.shopp.ApiHandler.DataFetcherCallBack;
import com.ramez.shopp.Classes.Constants;
import com.ramez.shopp.Classes.UtilityApp;
import com.ramez.shopp.Models.LocalModel;
import com.ramez.shopp.Models.ProductModel;
import com.ramez.shopp.Models.Recipe;
import com.ramez.shopp.Models.ResultAPIModel;
import com.ramez.shopp.R;
import com.ramez.shopp.databinding.RowRecipeBinding;

import java.util.ArrayList;
import java.util.List;

public class RecipeAdapter extends RecyclerView.Adapter<RecipeAdapter.ViewHolder> implements SuggestedProductAdapter.OnItemClick {
    private Activity activity;
    private List<Recipe> list;
    private DataFetcherCallBack dataFetcherCallBack;
    SuggestedProductAdapter adapter;
    int country_id, city_id, userId = 0;
    LocalModel localModel;


    public RecipeAdapter(Activity activity, List<Recipe> list, int selectedPosition, DataFetcherCallBack dataFetcherCallBack) {
        this.activity = activity;
        this.list = list;
        this.dataFetcherCallBack = dataFetcherCallBack;

        localModel = UtilityApp.getLocalData() != null ? UtilityApp.getLocalData() : UtilityApp.getDefaultLocalData(activity);
        country_id = localModel.getCountryId();
        city_id = Integer.parseInt(localModel.getCityId());

        if (UtilityApp.getUserData() != null) {
            userId = UtilityApp.getUserData().getId();

        }
    }

    @NonNull
    @Override

    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RowRecipeBinding itemView = RowRecipeBinding.inflate(LayoutInflater.from(activity), parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        Recipe recipe = list.get(position);
        holder.binding.nameTxt.setText(recipe.getDescription());

        holder.binding.selectTxt.setText(activity.getString(R.string.fa_check));
        holder.binding.selectTxt.setTextColor(ContextCompat.getColor(activity, R.color.green));

        if (recipe.isOpen) {
            // to open and change header icon
            holder.binding.toggleBut.setText(activity.getString(R.string.fa_angle_up));
            holder.binding.selectTxt.setVisibility(View.VISIBLE);

            if (recipe.isLoaded) {
                // to check if products loaded before
                adapter = new SuggestedProductAdapter(activity, recipe.productsList, this, 0);
                holder.binding.rv.setAdapter(adapter);

                holder.binding.progressBar1.setVisibility(View.GONE);
                holder.binding.noProductsTv.setVisibility(View.GONE);
                holder.binding.rv.setVisibility(View.VISIBLE);

            } else {
                // if products not loaded
                getProductRecipeList(holder.binding, recipe, country_id, city_id, String.valueOf(userId));
            }

        } else {
            holder.binding.toggleBut.setText(activity.getString(R.string.fa_angle_down));
            holder.binding.selectTxt.setVisibility(View.GONE);
            holder.binding.rv.setVisibility(View.GONE);
        }

    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public void onItemClicked(int position, ProductModel productModel) {

    }


    public class ViewHolder extends RecyclerView.ViewHolder {

        RowRecipeBinding binding;

        public ViewHolder(RowRecipeBinding view) {
            super(view.getRoot());
            binding = view;


            LinearLayoutManager llm = new LinearLayoutManager(activity, RecyclerView.HORIZONTAL, false);
            binding.rv.setLayoutManager(llm);

            binding.rv.setHasFixedSize(true);
            binding.rv.setItemAnimator(null);


            itemView.setOnClickListener(v -> {

                Recipe recipe = list.get(getBindingAdapterPosition());
                recipe.isOpen = !recipe.isOpen;
                notifyItemChanged(getBindingAdapterPosition());


            });

        }

    }


    private void getProductRecipeList(RowRecipeBinding binding, Recipe recipe, int country_id, int city_id, String user_id) {
//        productList.clear();
        binding.progressBar1.setVisibility(View.VISIBLE);
        binding.rv.setVisibility(View.GONE);
        binding.noProductsTv.setVisibility(View.GONE);

        new DataFeacher(false, (obj, func, IsSuccess) -> {
            binding.progressBar1.setVisibility(View.GONE);

            ResultAPIModel<ArrayList<ProductModel>> result = (ResultAPIModel<ArrayList<ProductModel>>) obj;

            if (IsSuccess) {
                if (result.data != null && result.data.size() > 0) {

                    recipe.productsList = result.data;
                    recipe.isLoaded = true;

                    SuggestedProductAdapter adapter = new SuggestedProductAdapter(activity, recipe.productsList, this, 0);
                    binding.rv.setAdapter(adapter);
                    binding.rv.setVisibility(View.VISIBLE);
                    binding.noProductsTv.setVisibility(View.GONE);
                } else {
                    binding.rv.setVisibility(View.GONE);
                    binding.noProductsTv.setVisibility(View.VISIBLE);

                }

            }


        }).getProductRecipeList(recipe.getId(), country_id, city_id, user_id, 0, 10);
    }


}

File: app/src/main/java/com/amour/shop/adapter/RecipeAdapter.kt
package com.amour.shop.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.amour.shop.ApiHandler.DataFeacher
import com.amour.shop.ApiHandler.DataFetcherCallBack
import com.amour.shop.BuildConfig
import com.amour.shop.Models.LocalModel
import com.amour.shop.Models.ProductModel
import com.amour.shop.Models.Recipe
import com.amour.shop.Models.ResultAPIModel
import com.amour.shop.R
import com.amour.shop.classes.Constants
import com.amour.shop.classes.UtilityApp
import com.amour.shop.classes.UtilityApp.getLocalData
import com.amour.shop.classes.UtilityApp.getUserData
import com.amour.shop.databinding.RowRecipeBinding


class RecipeAdapter(
    private val activity: Context,
    private val list: List<Recipe>?,
    selectedPosition: Int,
    private val dataFetcherCallBack: DataFetcherCallBack
) :
        RecyclerView.Adapter<RecipeAdapter.ViewHolder>(),SuggestedProductAdapter.OnItemClick {
    var adapter: SuggestedProductAdapter? = null
    var countryId: Int
    var cityId: Int
    var userId = 0
    var localModel: LocalModel? = getLocalData()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = RowRecipeBinding.inflate(LayoutInflater.from(activity), parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        localModel= getLocalData()
        val recipe = list?.get(position)
        holder.binding.nameTxt.text = recipe?.description
        holder.binding.selectTxt.text = activity.getString(R.string.fa_check)
        holder.binding.selectTxt.setTextColor(ContextCompat.getColor(activity, R.color.green))
        if (recipe?.isOpen == true) {
            // to open and change header icon
            holder.binding.toggleBut.text = activity.getString(R.string.fa_angle_up)
            holder.binding.selectTxt.visibility = View.VISIBLE
            if (recipe.isLoaded) {
                // to check if products loaded before
                adapter = SuggestedProductAdapter(activity, recipe.productsList, this, 0)
                holder.binding.rv.adapter = adapter
                holder.binding.progressBar1.visibility = View.GONE
                holder.binding.noProductsTv.visibility = View.GONE
                holder.binding.rv.visibility = View.VISIBLE
            } else {
                // if products not loaded
                getProductRecipeList(holder.binding, recipe, countryId, cityId, userId.toString())
            }
        } else {
            holder.binding.toggleBut.text = activity.getString(R.string.fa_angle_down)
            holder.binding.selectTxt.visibility = View.GONE
            holder.binding.rv.visibility = View.GONE
        }
    }

    override fun getItemCount(): Int {
        return list?.size?:0
    }

    override fun onItemClicked(position: Int, productModel: ProductModel?) {}
    inner class ViewHolder(var binding: RowRecipeBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        init {
            val llm = LinearLayoutManager(activity, RecyclerView.HORIZONTAL, false)
            binding.rv.layoutManager = llm
            binding.rv.setHasFixedSize(true)
            binding.rv.itemAnimator = null
            itemView.setOnClickListener { v: View? ->
                val recipe = list?.get(bindingAdapterPosition)
                recipe?.isOpen = !recipe?.isOpen!!
                notifyItemChanged(bindingAdapterPosition)
            }
        }
    }

    private fun getProductRecipeList(
        binding: RowRecipeBinding,
        recipe: Recipe,
        country_id: Int,
        city_id: Int,
        user_id: String
    ) {
//        productList.clear();
        binding.progressBar1.visibility = View.VISIBLE
        binding.rv.visibility = View.GONE
        binding.noProductsTv.visibility = View.GONE
        DataFeacher(false, object :DataFetcherCallBack {
            override fun Result(obj: Any?, func: String?, IsSuccess: Boolean) {
                binding.progressBar1.visibility = View.GONE
                val result =
                    obj as ResultAPIModel<ArrayList<ProductModel>?>
                if (IsSuccess) {
                    if (result.data != null && result.data!!.size > 0) {
                        recipe.productsList = result.data
                        recipe.isLoaded = true
                        val adapter =
                            SuggestedProductAdapter(activity, recipe.productsList,this@RecipeAdapter, 0)
                        binding.rv.adapter = adapter
                        binding.rv.visibility = View.VISIBLE
                        binding.noProductsTv.visibility = View.GONE
                    } else {
                        binding.rv.visibility = View.GONE
                        binding.noProductsTv.visibility = View.VISIBLE
                    }
                }            }

        }).getProductRecipeList(recipe.id, country_id, city_id, user_id, 0, 10)
    }

    init {
        countryId = localModel!!.countryId
        cityId = localModel!!.cityId.toInt()
        if (getUserData() != null) {
            userId = getUserData()!!.id
        }
    }
}